File: java/polyglot-utils-java/src/main/java/dev/jdwilliams/utils/math/MathUtils.java
package dev.jdwilliams.utils.math;

public final class MathUtils {

    private MathUtils() {
        // Prevent instantiation
    }

    /**
     *  Returns the greatest common divisor (GCD) of two integers using the
     *  Euclidean algorithm.
     */
    public static int gcd(int a, int b) {
        a = Math.abs(a);
        b = Math.abs(b);
        int max = Math.max(a, b);
        int min = Math.min(a, b);
        while (min != 0) {
            int temp = min;
            min = max % min;
            max = temp;
        }
        return max;
    }

    /**
     *  Returns the least common multiple (LCM) of two integers.
     */
    public static int lcm(int a, int b) {
        if (a == 0 || b == 0) {
            return 0;
        }
        return Math.abs(a * b) / gcd(a, b);
    }

    /**
     *  Returns true if a number is prime .
     */
    public static boolean isPrime(int n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        for (int i = 5; i * i <= n; i+=6) {
            if (n % i == 0 || n % (i + 2) == 0) return false;
        }
        return true;
    }

    /**
     *  Returns the factorial of a non-negative integer.
     */
    public static long factorial(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("n must be non-negative");
        }
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }
}

File: python/src/jdwilliams_utils/math/math_utils.py
def factorial(n: int) -> int:
    """Compute the factorial of a non-negative integer.

    Args:
        n (int): The integer to compute the factorial for.

    Returns:
        int: The factorial of n.

    Raises:
        ValueError: If n is negative.
    """
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

def gcd(a: int, b: int) -> int:
    """Compute the greatest common divisor of two integers using the Euclidean algorithm.

    Args:
        a (int): First integer.
        b (int): Second integer.

    Returns:
        int: The greatest common divisor of a and b.
    """
    a = abs(a)
    b = abs(b)
    max_val = a if a > b else b
    min_val = a if a < b else b
    while min_val != 0:
        max_val, min_val = min_val, max_val % min_val
    return max_val

def is_prime(n: int) -> bool:
    """
    Return True if the given integer is a prime number.

    A prime number is an integer greater than 1 that has no positive
    divisors other than 1 and itself.
    """
    if n < 2:
        return False
    
    if n <= 3:
        return True
    
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    limit = int(n**0.5) + 1
    for i in range(5, limit, 6):
        if n % i == 0 or n % (i + 2) == 0:
            return False
    return True

def lcm(a: int, b: int) -> int:
    """Compute the least common multiple of two integers.

    Args:
        a (int): First integer.
        b (int): Second integer.

    Returns:
        int: The least common multiple of a and b.
    """
    if a == 0 or b == 0:
        return 0
    return abs(a * b) // gcd(a, b)

